File: Week_02-哈希表-集合-树-图/350.两个数组的交集-ii.py
#
# @lc app=leetcode.cn id=350 lang=python
#
# [350] 两个数组的交集 II
#

# @lc code=start
class Solution(object):
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        #数组放入字典再查询，注意字典内要对元素计数
        n1 = len(nums1)
        n2 = len(nums2)
        dictable1 = dict()
        ans = list()
        for i in range(n1):
            if nums1[i] in dictable1:
                dictable1[nums1[i]] += 1
            else:
                dictable1[nums1[i]] = 1
        for j in range(n2):
            if nums2[j] in dictable1 and dictable1[nums2[j]] > 0:
                ans.append(nums2[j])
                dictable1[nums2[j]] -= 1
        return ans


# @lc code=end

File: Week_02-哈希表-集合-树-图/350.两个数组的交集-ii.cpp
/*
 * @lc app=leetcode.cn id=350 lang=cpp
 *
 * [350] 两个数组的交集 II
 */

// @lc code=start
#include<vector>
#include<unordered_map>
using namespace std;
class Solution {
public:
        /*
        方法一：hash_map
        思路：属于比较两组数据内容关系的题目，可以用hash_map表示其中一组数据的元素分布情况，
        用另外一个数组逐元素查询hash_map，统计两者关系。
        步骤：
        1.如果nums1元素个数大于nums2，则交换数组元素；
        2.把nums1数组放入hash_map，value为相应元素（key）在数组中的个数；
        3.遍历nums2，检查是否在m中存在：
           如果存在，且计数（value）为正，则
               把元素拷贝到nums1[k]，且k++;
               m中相应元素的计数减1
        4.返回nums1的前k个元素

        时间复杂度O(m+n)
        空间复杂度0(min(n,m))，因为第一步取较小素组存入hash_map
        */
    #if 0
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {

        if (nums1.size() > nums2.size()) {
            return intersect(nums2, nums1);  //调换实参顺序的操作
        }
        unordered_map<int, int> m;
        for (auto n : nums1) {
            m[n]++;        //数组放入hash_map的操作
        }
        int k = 0;
        for (auto n : nums2) {
            auto it = m.find(n);
            if (it != end(m) && --it->second >= 0) {    //先减再比较大小，也可以写成: it->second-- > 0
                nums1[k++] = n;
            }
        }
        return vector(begin(nums1), begin(nums1) + k);
    }
    #endif

    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() > nums2.size())
            return intersect(nums2, nums1);
        unordered_map<int, int> m;
        for (auto n : nums1) {
            m[n]++;
        }
        int k = 0;
        for (auto n : nums2) {
            auto it = m.find(n);
            if (it != end(m) && --it->second >=0) {
                nums1[k++] = n;
            }
        }
        return vector(begin(nums1), begin(nums1) + k);
    }








};
// @lc code=end

